### ISSUE TITLE
Script Bundling gmail.js with React/Webpack

### ISSUE THREAD

**Issue by tangentToNothing** [KartikTalwar/gmail.js#644]
Love this repo, thanks for maintaining it over the years!

I am building a chrome extension with React and Webpack for bundling. Does anyone know of examples of bundling gmail.js in this way and making it available in a content script? I have tried two approaches:

1. Direct injecting in content scripts:
- copy `gmail.js` to `dist/` and instruct manifest to include `gmail.js` alongside my bundled content script(s)

```js
// snippet of content-script

chrome.runtime.onMessage.addListener(function(request, sender, callback) {
  // snip
    if (request.message == 'load') {
        main();
    }
// snip
});

// snip

function main() {
    var gmail = Gmail();
    console.log("finished loading");
    var user_email = gmail.get.user_email();
    console.log(user_email);
}

```

*results:*
- `gmail.get.user_email()` yields user email! DOM scraping elements work great
- `gmail.observe.on(<anything>)` all fail to bind and I can't call any functions bound to the events

2. `npm install` and bundling
- `require` gmail.js

```js
// snippet of content-script

const gmailjs = require("gmail-js");

const gmail = new gmailjs.Gmail();

gmail.observe.on("load", () => { 

    gmail.observe.on("view_email", (domEmail) => { 
        var user_email = gmail.get.user_email();
    
        console.log(user_email);
        console.log(domEmail);

        console.log("Looking at email:", domEmail);
            const emailData = gmail.new.get.email_data(domEmail);
            console.log("Email data:", emailData);
    });    
});
```

*results:*
- `gmail.observe.on(<anything>)` bind functions work great! In the above example, I receive the `ID` of the email I am currently viewing
- most or all DOM scraping things fail, including the `gmail.get.user_email()` in the example above. No errors, just `undefined` in the variable.

Any help would be greatly appreciated!

**Comment by josteink** [comment]
As you've noticed, when you bundle Gmail.js with a (extension using a) large framework like React, things no longer work as smoothly as in the simple case.

This is because Gmail.js needs to get *early* into the GMail loading process to handle all events, and obtain all the properties it needs.

And when you create a giga-bundle with all batteries and stuff included in a single, big content-script, that content-script will take longer to load, longer to parse, longer to init. And then chances are that GmailJS will miss the boat. Ooops!

The solution to this is to create 2 content-scripts:

* An gmail.js preloader script which contains *nothing* but gmail.js, initializes it, and puts it *somewhere* in the `window`-object. This payload will be small and load fast.
* Your React-extension, which can be slow to load, parse and init, and which picks up your GmailJS instance from where you left it in the `window`-object.

And if you do that, you will *probably* see your problems go away.

**Comment by tangentToNothing** [comment]
I may be missing something obvious... In the preloader, any advice on ensuring jQuery is already there to initialize a new  `Gmail` object? I placed gmail.js directly in the extension and, for testing purposes, appended a few lines to add it to the window:

```js
//end of gmail.js
window._gmailjs = window._gmailjs || new Gmail();
```
--> `gmail.get.user_email()` = `undefined`

```js
// end of gmail.js
const jQuery = require("jquery");
window._gmailjs = window._gmailjs || new Gmail(jQuery);
```
--> `Uncaught ReferenceError: require is not defined
    at gmail.js:4266`

In both cases I followed the example in the extension repo to grab the `_gmailjs` object from the window in the second content script:

```js
// content_script.js
const loaderId = setInterval(() => {
    if (!window._gmailjs) {
        return;
    }

    clearInterval(loaderId);
    startExtension(window._gmailjs);
}, 100);

function startExtension(gmail) {
    window.gmail = gmail;

    gmail.observe.on("load", () => {
        const userEmail = gmail.get.user_email();
        console.log(userEmail);
    });
}
```

**Comment by josteink** [comment]
> I may be missing something obvious... In the preloader, any advice on ensuring jQuery is already there to initialize a new Gmail object? 

Oh sorry. I forgot. Yes, you do need to also bundle jQuery with Gmail.js in the preloader-script.

If you do that, everything should work just fine. Here is a (somewhat modified) content-script I inject at the same time as my "full" content-script (note that this is typescript):

````typescript

// gmail.js needs to be loaded as early as possible to be able to intercept
// embedded email-data in the gmail HTML!
//
// To do that we:
// - just load it
// - make sure it's initialized
// - and do nothing else!
//
// Let the "big" extension bundle load separately!

const GmailFactory = require("gmail-js");
import * as jQuery from "jquery";

let props = (window as any).MyVendorProps;
props.Vendor = {};
props.Vendor.Gmail = new GmailFactory.Gmail(jQuery) as Gmail;
````

**Comment by tangentToNothing** [comment]
I'm not quite familiar with TS, but I adjusted webpack config to accept TS for my preloader getting this error:

```ts
// relevant part of pre_load script
let props = (window as any).MyVendorProps;
props.Vendor = {};
props.Vendor.Gmail = new GmailFactory.Gmail(jQuery);
```

error:

``` 
Uncaught TypeError: Cannot set property 'Vendor' of undefined
    at Module../src/pages/PreLoad/index.ts (pre_load.bundle.js:15675)
    at __webpack_require__ (pre_load.bundle.js:15908)
    at pre_load.bundle.js:16887
    at pre_load.bundle.js:16890 
```

Have a JS version? Thank you so much for the help so far!

**Comment by tangentToNothing** [comment]
gave this a try as well, no luck: 

```ts
(<any>window)._gmailjs = new GmailFactory.Gmail(jQuery);
```

Webpack was tossing errors on the TS `as Gmail` when setting `Gmail` on the Vendor Props:
```
      TS2304: Cannot find name 'Gmail'.
```

**Comment by josteink** [comment]
Yeah. You may have to initialize the window-object too. My bad. 

Just try to get the preloader script working first (remember it must be injected!), and the content script injecting it can't be big either.

**Comment by Avivonz** [comment]
Any resolution on this one?
Facing the exact same issue - 
Even when simply bundling this file without anything - 
`const GmailFactory = require("gmail-js");
const jQuery = require("jquery");

window._gmailjs = window._gmailjs || new GmailFactory.Gmail(jQuery);

gmail.observe.on("load", () => {
    console.log(gmail);
    console.log(gmail.get.user_email());
})`
The gmail object is instantiated (observing works as well) but the user_email comes as undefined.
The manifest.json is as follows as well so should be on point:
`{
      "matches": [
        "*://mail.google.com/*"
      ],
      "js": [
        "gmailPreloader.bundle.js"
      ],
      "run_at": "document_start"
    }`

**Comment by josteink** [comment]
Can you post a complete minimal proof-of-concept repo?

A web-extension contains many parts, and a GmailJS web-extension even more (since it needs preloading and injecting into non-contentscript DOM).

It's almost impossible to an effective diagnosis without looking at the system as a whole.

**Comment by josteink** [comment]
To be clear: To me (based on your code so far, and the results it provides) it sounds like you're not injecting the script bundle into the non-contentscript DOM, but I cannot verify that without your full POC.

**Comment by Avivonz** [comment]
> To be clear: To me (based on your code so far, and the results it provides) it sounds like you're not injecting the script bundle into the non-contentscript DOM, but I cannot verify that without your full POC.

Sure, even though I'm not quite sure that's the problem and what you're referring to as non-contentscript DOM as from to my understanding all that's required is to inject gmail.js into gmail's context.
To prove my point, the following code produces the following logs:
``
gmail.observe.on("view_email", async (domEmail: any) => {
	console.log("Looking at email:", domEmail);
	const emailData = gmail.new.get.email_data(domEmail);
	console.log("Email data:", emailData);
	const response = await client.message('/request/test', {
		"data": emailData
	});
	console.log("Got response!", response.data);
});
```
![image](https://user-images.githubusercontent.com/9980033/112192383-93fa8900-8c0f-11eb-9434-3eb1fc0c1772.png)

The element is the correct element upon inspection, so it just fails to extract the data for some reason, same for current email, how would you like me to attach the poc repo? (Wanna move a medium outside of here where we resolve the issue then come back here?)

**Comment by giannif** [comment]
@josteink thanks for this library, it saved us as we had to move away from InboxSDK in a hurry because they all of a sudden wanted to bill us a very large monthly rate 

We use React and bundle with [Parcel](https://parceljs.org/). I haven't seen any issues in development and am wondering if I need to code split a bootstrapper as you recommend. 

We're only using the compose window feature. 
```tsx
gmail.observe.on('compose', (compose: Compose) => {...
```

Maybe that doesn't require code splitting? 

We track runtime errors in TrackJS and the only suspicious one so far is `target` here is null:
<img width="670" alt="Screen Shot 2021-04-20 at 9 15 16 AM" src="https://user-images.githubusercontent.com/448767/115319974-cb3c5700-a135-11eb-9ae0-9a0c5d1b4fa7.png">

This happened only a few times. 

Would there be an obvious error if gmailjs is not _getting in early_ enough? I'd obviously like to avoid the code splitting but could get it on our roadmap if we think we need it. Any guidance would be appreciated 🙏

**Comment by josteink** [comment]
Hey there and thanks for the kind words.

Looking at the piece of code you have highlighted, you can see 4 codelines up that there have been logging/debugging-statements related to these exact pieces of code already, so clearly this part of the code is somewhat touchy.

Looking at the rest of the code, the only place it's getting invoked is here: 

![image](https://user-images.githubusercontent.com/411338/115563558-f6e74a80-a2b7-11eb-8f2c-61ef26ddc6f6.png)

Based on that, it seems weird to me then that `target` can actually be `null`! Though this is not really my code and I don't know it intimately enough to know what to expect or not expect here.

That said, I know that timing-related errors are really annoying to work with, debug and generally have to handle in any form or way.

I don't know IF this is a timing-related error, but it would certainly help eliminate that possibility completely by having gmail.js load up front as a separate bundle before loading the "heavy" stuff.

If you still encounter issues after that, at least you have one less thing to worry about :smile:

**Comment by SimonBiggs** [comment]
Hi @josteink,

Thank you so very much for building out this library :slightly_smiling_face:. I'm trying to find my feet with gmail.js when used in combination with React and Typescript.

I've been trying to pull in email.js into the following boilerplate library:
https://github.com/akoskm/vite-react-tailwindcss-browser-extension

But with no success as of yet. Do you happen to know of an example online where someone has successfully bundled gmail.js along with React + Typescript? I have been looking around and unfortunately, thus far, have come up cold :slightly_frowning_face:

**Comment by josteink** [comment]
I mean... I haven't *seen* anyone do this, but I don't see why it can't be done?

* Browser-extensions using Gmail.js and Typescript works fine.
* React is a normal JS library, and you should be able to bundle it just fine.
* Browser-extensions built with React [even have a template](https://dev.to/iyashsoni/building-web-extensions-with-reactjs-from-0-to-publish-54no) (or so a quick internet search tells me).

You just need to combine the build-technical requirements from both of those worlds into 1 extension:

* Typescript + TSX compilation and bundling.
* Gmail.js script injected into regular DOM from content-script
* Pick up your injected Gmail.js instance in another injected script

Basically, just build your React-based extension and make sure it gets injected into the page, rather than run as a content-script, and you should almost already be there.

Since React-development has a kinda of "complicated" build-procedure, I would probably recommend start on the React side, and then add the few things required by Gmail.js to that process.

If you haven't already, there's a very minimal POC or how to use Gmail.js without *any* other distractions  in the boilerplate-repo: https://github.com/josteink/gmailjs-node-boilerplate

**Comment by josteink** [comment]
> Here is where I am up to: https://github.com/assistancechat/assistance.chat/tree/1aff495/src/typescript/browser-extensions/assistance

I don't see anything fundamentally wrong with that. It seems to have most of the right bits in the right place.

It does work now, right?

> Also, if I had some PRs that included extra typing information would that be welcome?

PRs are always welcome.

**Comment by SimonBiggs** [comment]
Yup, it does work. I should have said that. I was just aiming to leave some breadcrumbs for others to follow in the future.

**Comment by SimonBiggs** [comment]
> PRs are always welcome.

So, I've been chasing various options in the AI space. And for now, given Microsoft's own AI browser "extension" I am not going to continue down this browser extension path just for now.

**Comment by josteink** [comment]
No hard feelings. Whatever works for you 😋
